### PR TITLE
Make SET command two-phase-commit aware

### PR DESCRIPTION
**Pull request by pengzhout**
The flag two-phase-commit is used to be false for SET commands which
make SET commands cannot be rolled back in a transaction which does
not match the upstream, this commit fixes this defect.

### CODE REVIEW

File: src/test/regress/output/dispatch.source [hunk 1]
```diff
@@ -773,5 +773,48 @@ select count(*) from dispatch_foo, dispatch_bar where exists (select dispatch_fo
...
+
+select gp_inject_fault('variable_set', 'error', 2);
+NOTICE:  Success:  (seg0 10.153.101.106:25432 pid=458627)
+ gp_inject_fault 
+-----------------
+ t
+(1 row)
+
+-- set command should fail, default_with_oids should still be true
+set default_with_oids to false;
```
Comment (paul-guo-) [inline src/test/regress/output/dispatch.source]: default_with_oids was deprecated in pg upstream and it is very possible to be removed in gpdb code also. Please test using another GUC.

Comment (pengzhout) [reply src/test/regress/output/dispatch.source]: Good suggestion, all right, let's use another GUC and a new way to test this, please refer the latest commit pushed.

**Comment by hlinnaka** [comment]
I don't understand why SET needs two-phase commit. It doesn't change anything on disk.

**Comment by hlinnaka** [comment]
> > I don't understand why SET needs two-phase commit. It doesn't change anything on disk.
> 
> Take GUC like search_path, default_tablespace, temp_tablespace, default_with_oids or gp_default_storage_options for example, although it doesn't change anything on disk, it might affect the property of tables and the paths to look up tables or files. if gp_default_storage_options is inconsistent between segments, when creating a table, some segments may create appendonly column-store and others may create row-store for a same table.

Yeah, we need to make sure that the GUCs are consistent in all the QE processes. But doing two-phase commit seems like a very strong-armed way of doing it. And it doesn't handle all the cases, anyway: new QE processes need to set them according to the QD's current values, for example. We have code to handle many cases already, so it doesn't seem appropriate to suddenly start requiring two-phase commit in all cases.

The comment is not explaining the issue in enough detail, at the very least. Two-phase commit is usually concerned with making sure that persistent on-disk changes are committed in an all-or-nothing fashion in all the segments. If we're repurposing it for something else, it needs to be explained in much more detail.

**Comment by hlinnaka** [comment]
This reminds me that https://github.com/greenplum-db/gpdb/issues/685 is also still not fixed.

I think we should change our approach, wrt. keeping GUCs in sync in all the processes. Instead of immediately dispatching the SET command, I think we should have a per-connection flag, to indicate whether the GUCs in that QE process are in sync with the QD process. If the GUCs become out of sync for any reason, for example because the user issues a SET command, or because the connection was just established and the GUCs haven't been set yet, clear the flag. Check the flag before dispatching any command, and issue the required SETs at that point, if needed.

And instead of trying to keep track of which GUCs are in sync, dispatch all the GUCs that are different from the default values.

**Comment by pengzhout** [comment]
> Yeah, we need to make sure that the GUCs are consistent in all the QE processes. But doing two-phase commit seems like a very heavy-handed way of doing it. And it doesn't handle all the cases, anyway: new QE processes need to set them according to the QD's current values, for example. We have code to handle many cases already, so it doesn't seem appropriate to suddenly start requiring two-phase commit in all cases.

Hmm, got your points, two-phase-commit is too heavy, I'm wondering if it is reasonable to clear out all QEs if a SET command got errors, so GUCs will be synced to newly created QEs later, the only problem is the session will be reset if it has temp table.

> 
> The comment is not explaining the issue in enough detail, at the very least. Two-phase commit is usually concerned with making sure that persistent on-disk changes are committed in an all-or-nothing fashion in all the segments. If we're repurposing it for something else, it needs to be explained in much more detail.

Too bad not adding more detail comments, will improve it.

>I think we should change our approach, wrt. keeping GUCs in sync in all the processes. Instead of immediately dispatching the SET command, I think we should have a per-connection flag, to indicate whether the GUCs in that QE process are in sync with the QD process. If the GUCs become out of sync for any reason, for example because the user issues a SET command, or because the connection was just established and the GUCs haven't been set yet, clear the flag. Check the flag before dispatching any command, and issue the required SETs at that point, if needed.

Thanks for the approach, trying to understand it and taking a try.

**Comment by pengzhout** [comment]
> I think we should change our approach, wrt. keeping GUCs in sync in all the processes. Instead of immediately dispatching the SET command, I think we should have a per-connection flag, to indicate whether the GUCs in that QE process are in sync with the QD process. If the GUCs become out of sync for any reason, for example because the user issues a SET command, or because the connection was just established and the GUCs haven't been set yet, clear the flag. Check the flag before dispatching any command, and issue the required SETs at that point, if needed.

In this approach, we can remove the GUC related code within makeOptions() and treat it as a case to clear/set the flag. Are there any GUCs that must be set at the startup stage of processes? With this approach, we can postpone the SET command until the next query to save one libpq trip?

**Comment by ashwinstar** [comment]
@pengzhout based on discussion seems we should close this PR, evaluate alternate solution and open new PR for the same once its coded and ready.

**Comment by pengzhout** [comment]
> @pengzhout based on discussion seems we should close this PR, evaluate alternate solution and open new PR for the same once its coded and ready.

Yes, created a new issue https://github.com/greenplum-db/gpdb/issues/6610 to track this and close this.

**Pull request closed without merging**